src/DocumentDB.ChangeFeedProcessor/Estimator/RemainingWorkEstimator.cs

```diff
@@ -0,0 +1,162 @@
+﻿//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+//----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor.Estimator
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Linq;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.DataAccess;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Logging;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.PartitionManagement;
+    using Microsoft.Azure.Documents.Client;
+
+    internal class RemainingWorkEstimator : IRemainingWorkEstimator
+    {
+        private const string LSNPropertyName = "_lsn";
+        private static readonly ILog Logger = LogProvider.GetCurrentClassLogger();
+        private readonly IChangeFeedDocumentClient feedDocumentClient;
+        private readonly ILeaseManager leaseManager;
+        private readonly string collectionSelfLink;
+        private readonly int degreeOfParallelism;
+
+        public RemainingWorkEstimator(ILeaseManager leaseManager, IChangeFeedDocumentClient feedDocumentClient, string collectionSelfLink, int degreeOfParallelism)
+        {
+            if (leaseManager == null) throw new ArgumentNullException(nameof(leaseManager));
+            if (string.IsNullOrEmpty(collectionSelfLink)) throw new ArgumentNullException(nameof(collectionSelfLink));
+            if (feedDocumentClient == null) throw new ArgumentNullException(nameof(feedDocumentClient));
+            if (degreeOfParallelism < 1 || degreeOfParallelism > ServicePointManager.DefaultConnectionLimit) throw new ArgumentException("Degree of parallelism is out of range", nameof(degreeOfParallelism));
+
+            this.leaseManager = leaseManager;
+            this.collectionSelfLink = collectionSelfLink;
+            this.feedDocumentClient = feedDocumentClient;
+            this.degreeOfParallelism = degreeOfParallelism;
+        }
+
+        public async Task<long> GetEstimatedRemainingWork()
+        {
+            RemainingPartitionWork[] partitions = await this.GetEstimatedPartitionsRemainingWork();
+            if (partitions == null) return 0;
+
+            return partitions.Sum(partition => partition.RemainingWork);
+        }
+
+        public async Task<RemainingPartitionWork[]> GetEstimatedPartitionsRemainingWork()
+        {
+            IReadOnlyList<ILease> leases = await this.leaseManager.ListAllLeasesAsync().ConfigureAwait(false);
+            if (leases.Count == 0)
```

Comment:
In the original estimator, if there were no leases, then it'd return 1. This was so the user would at least run the Observer to check if there are indeed changes or not, and initialize the leases.

Reply:
This was covered by:

    if (!hasAtLeastOneLease)
    {
        return 1;
    }